### DISCUSSION TITLE
CS8509 false positive

### DISCUSSION THREAD

**Discussion by sheam** [dotnet/roslyn#49314] [Q&A]
**Version Used**:  

.Net Framework 4.7.2/C# 8.0

**Steps to Reproduce**:

`enum Position { Right, Left, Top }`

```
var position = Position.Last;
return position switch
{
      Position.Right=> "Right",
      Position.Left=> "Left",
      Position.Top => "Top ",
};
```

**Expected Behavior**:
No warning is issued with code as shown.
If I add a `Bottom` value to the `Position` enum, a CS8509 warning should be shown.

**Actual Behavior**:
CS8509 (about the switch not being exhaustive) warning is given unless I add a default case.

**Comments**
When I add `Bottom` to my enum, I want there to be a compile error or warning on the switch statement. If the point of CS8509 is not to provide this safety, is there another warning or way to do this?

**Comment by CyrusNajmabadi** [comment]
> is there another warning or way to do this?

Yes. This is an ideal place to use an analyzer

**Reply by sheam** [reply]
You mean I need a custom analyzer to make it work as expected?

**Reply by CyrusNajmabadi** [reply]
Yes. You would need a custom analyzer to add new errors to the language for code that is currently legal.

**Reply by sheam** [reply]
So why doesn't the language of the warning be changed to 'switch statement does not provide a default option'?  The current language is really confusing.

I am surprised a custom analyzer would be needed this seems like a common source of bugs.... I'l add a feature request.

**Reply by CyrusNajmabadi** [reply]
I don't see what's wrong with the language of the warning.  It seems very very clear and reasonable to me:

> The switch expression does not handle some values of its input type (it is not exhaustive) involving an unnamed enum value. For example, the pattern '(Position)3' is not covered.

It's showing how the switch doesn't cover all cases, and it even shows a specific example of a case that it doesn't match.

> I am surprised a custom analyzer would be needed

The language explicitly allows htis code.  So we cannot error on code which is explicitly legal.  

> I'l add a feature request.

I think you could just write the analyzer yourself :)   That's why we have them.  So that it's possible to add these sorts fo features without needing the language/ide to do it.  :)

**Comment by sheam** [comment]
I'll look into writing the analyzer, and could contrib it . THanks for the
feature request.

**Reply by CyrusNajmabadi** [reply]
You're welcome!
